# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

## Дополнительная часть (необязательная)

Выполнить задания 1-3 для всего набора данных.

# Обязательная часть

## Задание 1. Базовое изучение

In [1]:
import pandas as pd
import numpy as np

In [2]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [3]:
df =  pd.read_csv('horse_data.txt', names = ['surgery', 'age', 'hosp_no', 'rect_temp', 'pulse', 'resp_rate','extrem_temp', 'per_pulse', 'muc_memr', 'cap_refill', 'pain', 'peristalsis', 'abd_distension', 'nas_tube', 'nas_reflux','nas_refluxPH', 'rect-examination', 'abdomen', 'cell_volume', 'total_protein', 'abd_appearance', 'abd_total_protein', 'outcome', 'surgical_lesion', 'lesion1', 'lesion2', 'lesion3', 'cp_data'])

In [4]:
df.head()


surgery  age  hosp_no rect_temp pulse resp_rate extrem_temp per_pulse  \
0       2    1   530101     38.50    66        28           3         3   
1       1    1   534817      39.2    88        20           ?         ?   
2       2    1   530334     38.30    40        24           1         1   
3       1    9  5290409     39.10   164        84           4         1   
4       2    1   530255     37.30   104        35           ?         ?   

  muc_memr cap_refill  ... cell_volume total_protein abd_appearance  \
0        ?          2  ...       45.00          8.40              ?   
1        4          1  ...          50            85              2   
2        3          1  ...       33.00          6.70              ?   
3        6          2  ...       48.00          7.20              3   
4        6          2  ...       74.00          7.40              ?   

  abd_total_protein outcome surgical_lesion lesion1 lesion2 lesion3 cp_data  
0                 ?       2               2   11300       0       0       2  
1                 2       3               2    2208       0       0       2  
2                 ?       1               2       0       0       0       1  
3              5.30       2               1    2208       0       0       1  
4                 ?       2               2    4300       0       0       2  

[5 rows x 28 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   surgery            300 non-null    object
 1   age                300 non-null    int64 
 2   hosp_no            300 non-null    int64 
 3   rect_temp          300 non-null    object
 4   pulse              300 non-null    object
 5   resp_rate          300 non-null    object
 6   extrem_temp        300 non-null    object
 7   per_pulse          300 non-null    object
 8   muc_memr           300 non-null    object
 9   cap_refill         300 non-null    object
 10  pain               300 non-null    object
 11  peristalsis        300 non-null    object
 12  abd_distension     300 non-null    object
 13  nas_tube           300 non-null    object
 14  nas_reflux         300 non-null    object
 15  nas_refluxPH       300 non-null    object
 16  rect-examination   300 non-null    object
 1

In [6]:
df.replace('?', 'NaN', inplace=True)

In [7]:
df = df.astype('float')

In [8]:
df.describe()

surgery         age       hosp_no   rect_temp       pulse  \
count  299.000000  300.000000  3.000000e+02  240.000000  276.000000   
mean     1.397993    1.640000  1.085889e+06   38.167917   71.913043   
std      0.490305    2.173972  1.529801e+06    0.732289   28.630557   
min      1.000000    1.000000  5.184760e+05   35.400000   30.000000   
25%      1.000000    1.000000  5.289040e+05   37.800000   48.000000   
50%      1.000000    1.000000  5.303055e+05   38.200000   64.000000   
75%      2.000000    1.000000  5.347275e+05   38.500000   88.000000   
max      2.000000    9.000000  5.305629e+06   40.800000  184.000000   

        resp_rate  extrem_temp   per_pulse    muc_memr  cap_refill  ...  \
count  242.000000   244.000000  231.000000  253.000000  268.000000  ...   
mean    30.417355     2.348361    2.017316    2.853755    1.305970  ...   
std     17.642231     1.045054    1.042428    1.620294    0.477629  ...   
min      8.000000     1.000000    1.000000    1.000000    1.000000  ...   
25%     18.500000     1.000000    1.000000    1.000000    1.000000  ...   
50%     24.500000     3.000000    2.000000    3.000000    1.000000  ...   
75%     36.000000     3.000000    3.000000    4.000000    2.000000  ...   
max     96.000000     4.000000    4.000000    6.000000    3.000000  ...   

       cell_volume  total_protein  abd_appearance  abd_total_protein  \
count   271.000000     267.000000      135.000000         102.000000   
mean     46.295203      24.456929        2.037037           3.019608   
std      10.419335      27.475009        0.804905           1.968567   
min      23.000000       3.300000        1.000000           0.100000   
25%      38.000000       6.500000        1.000000           2.000000   
50%      45.000000       7.500000        2.000000           2.250000   
75%      52.000000      57.000000        3.000000           3.900000   
max      75.000000      89.000000        3.000000          10.100000   

          outcome  surgical_lesion       lesion1      lesion2      lesion3  \
count  299.000000       300.000000    300.000000   300.000000   300.000000   
mean     1.551839         1.363333   3657.880000    90.226667     7.363333   
std      0.737187         0.481763   5399.513513   649.569234   127.536674   
min      1.000000         1.000000      0.000000     0.000000     0.000000   
25%      1.000000         1.000000   2111.750000     0.000000     0.000000   
50%      1.000000         1.000000   2673.500000     0.000000     0.000000   
75%      2.000000         2.000000   3209.000000     0.000000     0.000000   
max      3.000000         2.000000  41110.000000  7111.000000  2209.000000   

          cp_data  
count  300.000000  
mean     1.670000  
std      0.470998  
min      1.000000  
25%      1.000000  
50%      2.000000  
75%      2.000000  
max      2.000000  

[8 rows x 28 columns]

In [9]:
df_dz1 = df[['surgery', 'age', 'rect_temp', 'pulse', 'pain','abd_appearance', 'total_protein', 'outcome']]

In [10]:
df_dz1.head()

surgery  age  rect_temp  pulse  pain  abd_appearance  total_protein  \
0      2.0  1.0       38.5   66.0   5.0             NaN            8.4   
1      1.0  1.0       39.2   88.0   3.0             2.0           85.0   
2      2.0  1.0       38.3   40.0   3.0             NaN            6.7   
3      1.0  9.0       39.1  164.0   2.0             3.0            7.2   
4      2.0  1.0       37.3  104.0   NaN             NaN            7.4   

   outcome  
0      2.0  
1      3.0  
2      1.0  
3      2.0  
4      2.0

In [11]:
metrics = df_dz1.describe()


In [12]:
mod = []

for col in df_dz1.columns:
    m = df_dz1[col].mode()[0]
    mod.append(m)

In [13]:
mod

[1.0, 1.0, 38.0, 48.0, 3.0, 2.0, 6.5, 1.0]

In [14]:
metrics.loc['moda'] = mod

In [16]:
metrics.loc['rang'] = metrics.loc['max'] - metrics.loc['min']

In [17]:
metrics.loc['lower_bound'] = metrics.loc['25%'] - (1.5 * ( metrics.loc['75%'] - metrics.loc['25%']))

In [18]:
metrics.loc['upper_bound'] = metrics.loc['75%'] + (1.5 * ( metrics.loc['75%'] - metrics.loc['25%']))

In [35]:
metrics

surgery         age   rect_temp       pulse       pain  \
count        299.000000  300.000000  240.000000  276.000000  245.00000   
mean           1.397993    1.640000   38.167917   71.913043    2.95102   
std            0.490305    2.173972    0.732289   28.630557    1.30794   
min            1.000000    1.000000   35.400000   30.000000    1.00000   
25%            1.000000    1.000000   37.800000   48.000000    2.00000   
50%            1.000000    1.000000   38.200000   64.000000    3.00000   
75%            2.000000    1.000000   38.500000   88.000000    4.00000   
max            2.000000    9.000000   40.800000  184.000000    5.00000   
moda           1.000000    1.000000   38.000000   48.000000    3.00000   
rang           1.000000    8.000000    5.400000  154.000000    4.00000   
lower_bound   -0.500000    1.000000   36.750000  -12.000000   -1.00000   
upper_bound    3.500000    1.000000   39.550000  148.000000    7.00000   

             abd_appearance  total_protein     outcome  
count            135.000000     267.000000  299.000000  
mean               2.037037      24.456929    1.551839  
std                0.804905      27.475009    0.737187  
min                1.000000       3.300000    1.000000  
25%                1.000000       6.500000    1.000000  
50%                2.000000       7.500000    1.000000  
75%                3.000000      57.000000    2.000000  
max                3.000000      89.000000    3.000000  
moda               2.000000       6.500000    1.000000  
rang               2.000000      85.700000    2.000000  
lower_bound       -2.000000     -69.250000   -0.500000  
upper_bound        6.000000     132.750000    3.500000

In [49]:
df_dz1.sort_values("pulse").head(100)

surgery  age  rect_temp  pulse  pain  abd_appearance  total_protein  \
232      1.0  1.0       38.5   30.0   NaN             NaN            7.7   
130      1.0  1.0       37.8   30.0   NaN             NaN            NaN   
242      1.0  1.0        NaN   36.0   5.0             NaN            5.9   
203      2.0  1.0       37.2   36.0   2.0             NaN            5.7   
276      1.0  1.0       38.0   36.0   4.0             2.0           75.0   
..       ...  ...        ...    ...   ...             ...            ...   
67       2.0  1.0       38.4   54.0   1.0             1.0            7.2   
284      2.0  1.0       38.5   54.0   3.0             2.0            6.8   
267      2.0  1.0       38.3   54.0   2.0             3.0            7.2   
85       2.0  1.0       38.0   54.0   3.0             NaN            6.2   
237      2.0  1.0       37.9   54.0   3.0             3.0           54.0   

     outcome  
232      1.0  
130      2.0  
242      2.0  
203      1.0  
276      3.0  
..       ...  
67       1.0  
284      1.0  
267      1.0  
85       1.0  
237      1.0  

[100 rows x 8 columns]

## Выводы


## Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [19]:
 for col in metrics.columns:
        if metrics[col].loc['min'] < metrics[col].loc['lower_bound'] or metrics[col].loc['max'] > metrics[col].loc['upper_bound']:
            print(f'{col} - выбросы есть' )

age - выбросы есть
rect_temp - выбросы есть
pulse - выбросы есть


In [20]:
df_dz1[df_dz1.age > 1]

surgery  age  rect_temp  pulse  pain  abd_appearance  total_protein  \
3        1.0  9.0       39.1  164.0   2.0             3.0            7.2   
9        2.0  9.0       38.3   90.0   5.0             1.0            6.2   
13       2.0  9.0       38.0   92.0   1.0             1.0            6.1   
16       1.0  9.0        NaN  128.0   4.0             3.0            7.8   
23       1.0  9.0       38.3  130.0   2.0             NaN           70.0   
39       1.0  9.0       39.2  146.0   NaN             NaN            NaN   
41       2.0  9.0       39.0  150.0   NaN             NaN            8.5   
55       1.0  9.0       38.6  160.0   3.0             2.0            NaN   
74       1.0  9.0        NaN    NaN   NaN             NaN            4.9   
75       1.0  9.0       39.7  100.0   2.0             2.0           57.0   
82       1.0  9.0       38.1  100.0   3.0             NaN            5.7   
103      1.0  9.0       38.0  140.0   3.0             NaN            5.3   
109      1.0  9.0       38.4   84.0   3.0             2.0            6.6   
135      1.0  9.0       38.1  136.0   5.0             2.0            4.9   
191      2.0  9.0        NaN  100.0   4.0             NaN            4.7   
212      1.0  9.0       38.9  120.0   3.0             1.0            6.3   
227      1.0  9.0        NaN    NaN   NaN             2.0            6.5   
229      1.0  9.0       38.5  120.0   NaN             1.0           54.0   
244      1.0  9.0       38.2  124.0   2.0             1.0            8.0   
252      1.0  9.0       38.8  124.0   2.0             3.0            7.6   
255      1.0  9.0       38.8  184.0   4.0             NaN            3.3   
262      2.0  9.0       38.6   88.0   NaN             NaN            5.9   
275      1.0  9.0       38.8  150.0   5.0             NaN            6.2   
287      2.0  9.0       39.5   84.0   NaN             NaN            5.0   

     outcome  
3        2.0  
9        1.0  
13       2.0  
16       2.0  
23       1.0  
39       2.0  
41       1.0  
55       2.0  
74       2.0  
75       3.0  
82       1.0  
103      1.0  
109      2.0  
135      2.0  
191      1.0  
212      1.0  
227      1.0  
229      1.0  
244      1.0  
252      2.0  
255      2.0  
262      1.0  
275      2.0  
287      1.0

In [21]:
df_dz1.age.value_counts()

1.0    276
9.0     24
Name: age, dtype: int64

In [22]:
pd.options.mode.chained_assignment = None  # default='warn'

In [23]:
df_dz1['age'].replace(9, np.NaN, inplace=True)


In [24]:
df_dz1.head(25)

surgery  age  rect_temp  pulse  pain  abd_appearance  total_protein  \
0       2.0  1.0       38.5   66.0   5.0             NaN            8.4   
1       1.0  1.0       39.2   88.0   3.0             2.0           85.0   
2       2.0  1.0       38.3   40.0   3.0             NaN            6.7   
3       1.0  NaN       39.1  164.0   2.0             3.0            7.2   
4       2.0  1.0       37.3  104.0   NaN             NaN            7.4   
5       2.0  1.0        NaN    NaN   2.0             NaN            NaN   
6       1.0  1.0       37.9   48.0   3.0             NaN            7.0   
7       1.0  1.0        NaN   60.0   NaN             NaN            8.3   
8       2.0  1.0        NaN   80.0   4.0             NaN            6.2   
9       2.0  NaN       38.3   90.0   5.0             1.0            6.2   
10      1.0  1.0       38.1   66.0   3.0             2.0            6.0   
11      2.0  1.0       39.1   72.0   2.0             NaN            7.8   
12      1.0  1.0       37.2   42.0   3.0             NaN            7.0   
13      2.0  NaN       38.0   92.0   1.0             1.0            6.1   
14      1.0  1.0       38.2   76.0   3.0             1.0           81.0   
15      1.0  1.0       37.6   96.0   5.0             NaN            6.8   
16      1.0  NaN        NaN  128.0   4.0             3.0            7.8   
17      2.0  1.0       37.5   48.0   NaN             NaN            NaN   
18      1.0  1.0       37.6   64.0   2.0             1.0            7.0   
19      2.0  1.0       39.4  110.0   NaN             NaN            8.7   
20      1.0  1.0       39.9   72.0   5.0             2.0            6.1   
21      2.0  1.0       38.4   48.0   1.0             NaN            6.8   
22      1.0  1.0       38.6   42.0   2.0             NaN            7.2   
23      1.0  NaN       38.3  130.0   2.0             NaN           70.0   
24      1.0  1.0       38.1   60.0   NaN             NaN           65.0   

    outcome  
0       2.0  
1       3.0  
2       1.0  
3       2.0  
4       2.0  
5       1.0  
6       1.0  
7       2.0  
8       3.0  
9       1.0  
10      1.0  
11      1.0  
12      1.0  
13      2.0  
14      1.0  
15      2.0  
16      2.0  
17      1.0  
18      1.0  
19      1.0  
20      1.0  
21      1.0  
22      1.0  
23      1.0  
24      1.0

#### Столбец Age выбросы выводы: 
Согласно описанию в этом столбце возможны только значения:  1 = Adult horse 2 = Young (< 6 months). Однако в данных есть только значения 1 и 9. У нас нет доказательств, чтобы допустить, что 9 это 2. В то же время значения 9 оставляют 8 % данных, что является значительным для их удаления. Решение : заменить 9 на NaN.


In [28]:
df_dz1.rect_temp.value_counts().sort_index()

35.4     1
36.0     1
36.1     1
36.4     1
36.5     2
36.6     1
36.8     1
36.9     1
37.0     2
37.1     3
37.2     7
37.3     6
37.4     3
37.5    12
37.6     7
37.7     7
37.8    17
37.9     7
38.0    26
38.1    12
38.2    16
38.3    18
38.4    11
38.5    19
38.6    12
38.7     7
38.8     6
38.9     4
39.0     4
39.1     2
39.2     5
39.3     4
39.4     3
39.5     4
39.6     1
39.7     1
39.9     1
40.0     1
40.3     2
40.8     1
Name: rect_temp, dtype: int64

In [29]:
df_dz1.pulse.value_counts().sort_index()

30.0      2
36.0      3
38.0      1
40.0     18
42.0     10
44.0     11
45.0      2
46.0      1
48.0     28
49.0      1
50.0      6
52.0     12
54.0      7
56.0      7
60.0     25
64.0      8
65.0      1
66.0      6
68.0      5
70.0      4
72.0     11
75.0      1
76.0      4
78.0      3
80.0      8
82.0      1
84.0      8
86.0      3
88.0     12
90.0      3
92.0      4
96.0      8
98.0      1
100.0    11
104.0     5
108.0     3
110.0     1
112.0     3
114.0     2
120.0    10
124.0     2
128.0     1
129.0     1
130.0     2
132.0     1
136.0     1
140.0     2
146.0     1
150.0     2
160.0     1
164.0     1
184.0     1
Name: pulse, dtype: int64

#### Столбцы rect_temp и pulse выбросы выводы: 
Выбросы распределены внешне равномерно и не выглядят аномально, поэтому считаю важным для анализа их оставить.


### Задание 3. Работа с пропусками

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [30]:
for col in df_dz1.columns:
    missing = df_dz1[col].isna().sum()
    print(f'{col} - {missing}')

surgery - 1
age - 24
rect_temp - 60
pulse - 24
pain - 55
abd_appearance - 165
total_protein - 33
outcome - 1


In [31]:
for col in df_dz1.columns:
    pct_missing = df_dz1[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.3%
age - 8.0%
rect_temp - 20.0%
pulse - 8.0%
pain - 18.3%
abd_appearance - 55.0%
total_protein - 11.0%
outcome - 0.3%


In [32]:
df_dz1[df_dz1['surgery'].isnull()]

surgery  age  rect_temp  pulse  pain  abd_appearance  total_protein  \
132      NaN  1.0       38.0   48.0   4.0             2.0           73.0   

     outcome  
132      NaN

#### Пропуски surgery и outcome. 
В этой строке нет 2 важных для данного датафрейма параметров: операция и результат. Т.к. это логические параметры, постановка средних значений  и наиболее часто встречающихся параметров может быть некорректно. Предлагаю удалить эту строку.

In [33]:
df_dz1.drop(132, inplace=True)


In [34]:
df_dz1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 299 entries, 0 to 299
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   surgery         299 non-null    float64
 1   age             275 non-null    float64
 2   rect_temp       239 non-null    float64
 3   pulse           275 non-null    float64
 4   pain            244 non-null    float64
 5   abd_appearance  134 non-null    float64
 6   total_protein   266 non-null    float64
 7   outcome         299 non-null    float64
dtypes: float64(8)
memory usage: 21.0 KB


#### Пропуски Age
 Здесь пропуски появились в результате преобразования некорретных данных и замена их на другие приведет к дальнейшему искажению, поэтому предлагаю их проигноприровать.
 

In [214]:
df_dz1[df_dz1['rect_temp'].isnull()]

surgery  age  rect_temp  pulse  pain  abd_appearance  total_protein  \
5        2.0  1.0        NaN    NaN   2.0             NaN            NaN   
7        1.0  1.0        NaN   60.0   NaN             NaN            8.3   
8        2.0  1.0        NaN   80.0   4.0             NaN            6.2   
16       1.0  NaN        NaN  128.0   4.0             3.0            7.8   
28       1.0  1.0        NaN    NaN   NaN             NaN            NaN   
34       1.0  1.0        NaN  100.0   5.0             NaN            6.6   
35       2.0  1.0        NaN  104.0   4.0             NaN            8.4   
40       1.0  1.0        NaN   88.0   5.0             3.0            6.5   
43       1.0  1.0        NaN  120.0   4.0             2.0           67.0   
45       2.0  1.0        NaN  120.0   5.0             3.0            6.5   
52       2.0  1.0        NaN    NaN   1.0             NaN            7.7   
56       1.0  1.0        NaN    NaN   NaN             NaN            6.7   
57       1.0  1.0        NaN   64.0   5.0             NaN            7.7   
58       1.0  1.0        NaN    NaN   5.0             3.0            5.9   
59       2.0  1.0        NaN   96.0   5.0             NaN            NaN   
65       1.0  1.0        NaN   60.0   3.0             2.0           72.0   
68       2.0  1.0        NaN   52.0   NaN             NaN            7.2   
74       1.0  NaN        NaN    NaN   NaN             NaN            4.9   
78       1.0  1.0        NaN    NaN   5.0             NaN            5.9   
86       1.0  1.0        NaN   88.0   5.0             3.0            7.7   
87       2.0  1.0        NaN   40.0   NaN             2.0            7.0   
93       2.0  1.0        NaN    NaN   5.0             NaN            6.5   
96       2.0  1.0        NaN   84.0   NaN             NaN            8.5   
101      1.0  1.0        NaN   60.0   5.0             NaN            NaN   
105      1.0  1.0        NaN   70.0   2.0             NaN            7.3   
110      1.0  1.0        NaN   70.0   2.0             NaN            7.5   
112      1.0  1.0        NaN   40.0   1.0             NaN           56.0   
115      2.0  1.0        NaN    NaN   3.0             NaN           70.0   
134      1.0  1.0        NaN  100.0   5.0             NaN            6.3   
136      1.0  1.0        NaN    NaN   5.0             NaN            5.9   
142      1.0  1.0        NaN  120.0   5.0             3.0            4.5   
145      2.0  1.0        NaN   80.0   2.0             NaN           80.0   
152      1.0  1.0        NaN  100.0   NaN             3.0           64.0   
159      2.0  1.0        NaN    NaN   2.0             2.0            7.5   
160      1.0  1.0        NaN    NaN   4.0             2.0           58.0   
167      2.0  1.0        NaN   44.0   3.0             NaN            6.3   
169      1.0  1.0        NaN   65.0   5.0             NaN            NaN   
173      1.0  1.0        NaN  104.0   2.0             2.0            8.6   
174      2.0  1.0        NaN    NaN   4.0             NaN            NaN   
175      1.0  1.0        NaN    NaN   NaN             NaN            NaN   
177      1.0  1.0        NaN   68.0   4.0             2.0            NaN   
178      1.0  1.0        NaN   60.0   5.0             3.0           70.0   
191      2.0  NaN        NaN  100.0   4.0             NaN            4.7   
193      1.0  1.0        NaN   60.0   4.0             NaN            7.6   
207      1.0  1.0        NaN    NaN   NaN             NaN            8.7   
216      2.0  1.0        NaN    NaN   NaN             NaN            NaN   
219      1.0  1.0        NaN   52.0   3.0             NaN            8.1   
222      1.0  1.0        NaN  112.0   2.0             3.0            6.3   
227      1.0  NaN        NaN    NaN   NaN             2.0            6.5   
242      1.0  1.0        NaN   36.0   5.0             NaN            5.9   
245      2.0  1.0        NaN   96.0   5.0             NaN            NaN   
253      2.0  1.0        NaN  112.0   5.0            

#### Выводы пропущенные значения.

 В данной точке я бы остановилась и проигнорировала все остальные пропуски, т.к. неизвестна задача исследования. В таком контексте замена, например, пропущенных значений температуры  на среднее может привести к неправильным выводам во время анализа.  
 В столбце  abd_appearance больше половины пропущенных значений (55%), гипотетически его можно удалить в случае, если эти показатели не имеют критического значения для анализа.
 
 В целях выполнения ДЗ сформируем датафрейм, в котором пропуски будут отсутствовать.
 
 1  Удаляем строки, в которых половина данных отсутствует.
 2  Решения по столбцам:
 
surgery - 0.3% - 1 строка, удаляем

age - 8.0% - заменим обратно на 'no data', чтобы видеть эту группу с недостаточными данными ( возможно это значение 2)

rect_temp - 20.0% - здесь мода - 38, медиана - 38,2 и среднее  арифм- 38,16 достаточно близки, пропуски заменим на среднее арифм. 

pulse - 8.0% - заменим на средн. арифметическое

pain - 18.3% - категориальная величина, меняем на моду - 3.

abd_appearance - 55.0% - т.к. больше половины данных отсутствует, удаляем столбец

total_protein - 11.0% - меняем на средн. арифметическое

outcome - 0.3% - 1 строка, удаляем
 
 

In [44]:
df_dz1[df_dz1['pulse'].isnull()]

surgery  age  rect_temp  pulse  pain  abd_appearance  total_protein  \
52       2.0  1.0        NaN    NaN   1.0             NaN            7.7   
58       1.0  1.0        NaN    NaN   5.0             3.0            5.9   
78       1.0  1.0        NaN    NaN   5.0             NaN            5.9   
83       1.0  1.0       38.0    NaN   5.0             NaN            7.8   
93       2.0  1.0        NaN    NaN   5.0             NaN            6.5   
115      2.0  1.0        NaN    NaN   3.0             NaN           70.0   
117      2.0  1.0       39.5    NaN   3.0             1.0            6.7   
126      1.0  1.0       38.7    NaN   5.0             2.0            6.5   
136      1.0  1.0        NaN    NaN   5.0             NaN            5.9   
151      1.0  1.0       39.3    NaN   4.0             3.0            NaN   
159      2.0  1.0        NaN    NaN   2.0             2.0            7.5   
160      1.0  1.0        NaN    NaN   4.0             2.0           58.0   
182      1.0  1.0       38.0    NaN   5.0             NaN            7.8   
204      1.0  1.0       39.2    NaN   4.0             1.0            6.6   

     outcome  
52       1.0  
58       2.0  
78       2.0  
83       2.0  
93       2.0  
115      1.0  
117      3.0  
126      1.0  
136      2.0  
151      2.0  
159      1.0  
160      1.0  
182      2.0  
204      1.0

In [42]:
df_dz1.dropna(thresh=5, inplace=True)

In [56]:
df_dz1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 287 entries, 0 to 299
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   surgery         287 non-null    float64
 1   age             287 non-null    object 
 2   rect_temp       287 non-null    float64
 3   pulse           273 non-null    float64
 4   pain            242 non-null    float64
 5   abd_appearance  133 non-null    float64
 6   total_protein   262 non-null    float64
 7   outcome         287 non-null    float64
dtypes: float64(7), object(1)
memory usage: 20.2+ KB


In [72]:
df_dz1.age.fillna('no_data', inplace=True)

In [55]:
df_dz1.rect_temp.fillna(df_dz1.rect_temp.mean(), inplace=True)

In [57]:
df_dz1.pulse.fillna(df_dz1.pulse.mean(), inplace=True)

In [59]:
df_dz1.pain.fillna(df_dz1.pain.mode()[0], inplace=True)

In [64]:
df_dz1.drop(['abd_appearance'], axis=1, inplace=True)

In [61]:
df_dz1.total_protein.fillna(df_dz1.total_protein.mean(), inplace=True)

In [77]:
df_dz1.describe()

surgery   rect_temp       pulse        pain  total_protein  \
count  287.000000  287.000000  287.000000  287.000000     287.000000   
mean     1.404181   38.164286   71.714286    2.954704      24.542748   
std      0.491590    0.666549   27.692853    1.203740      26.256546   
min      1.000000   35.400000   30.000000    1.000000       3.300000   
25%      1.000000   37.800000   48.000000    2.000000       6.600000   
50%      1.000000   38.164286   66.000000    3.000000       7.700000   
75%      2.000000   38.500000   88.000000    4.000000      54.500000   
max      2.000000   40.800000  184.000000    5.000000      89.000000   

          outcome  
count  287.000000  
mean     1.550523  
std      0.740953  
min      1.000000  
25%      1.000000  
50%      1.000000  
75%      2.000000  
max      3.000000

In [78]:
df_dz1

surgery      age  rect_temp  pulse  pain  total_protein  outcome
0        2.0        1  38.500000   66.0   5.0       8.400000      2.0
1        1.0        1  39.200000   88.0   3.0      85.000000      3.0
2        2.0        1  38.300000   40.0   3.0       6.700000      1.0
3        1.0  no_data  39.100000  164.0   2.0       7.200000      2.0
4        2.0        1  37.300000  104.0   3.0       7.400000      2.0
..       ...      ...        ...    ...   ...            ...      ...
295      1.0        1  38.164286  120.0   2.0      65.000000      3.0
296      2.0        1  37.200000   72.0   4.0      24.542748      3.0
297      1.0        1  37.500000   72.0   4.0       6.800000      2.0
298      1.0        1  36.500000  100.0   3.0       6.000000      1.0
299      1.0        1  37.200000   40.0   3.0      62.000000      3.0

[287 rows x 7 columns]